### ISSUE TITLE
ALOpsUpgrade - support Platform/Application upgrade in CD pipeline

### ISSUE THREAD

**Issue by fvet** [HodorNV/ALOps#233]
**Is your feature request related to a problem? Please describe.**
Can you provide some more insights on the ALOps roadmap for the upcoming releases? I would be really great if ALOps would be able to take care of '**Upgrading environments**' as part of the CD pipeline. This would actually enable us to complete the full CD process on DevOps **using ALOps only** (without the need of custom PS scripts).

**Describe the solution you'd like**
- We currently use ALOps in the CI pipeline to generate artifacts (app files)
- Artifacts are then deployed on target environment (hosted via OnPrem NST's with an external database) using CD pipelines.
- This is good for updates to our apps files that don't require a platform / application update. Once we switch to a higher minor / major BC version and include a new BC dependency (new platform / application), our CD pipelines cannot cope with this. 

Here's a high-level idea on what we would like to achieve **with ALOps**:

- (internal) Replace our OnPrem NST's by Docker containers. Still connected with external databases, since it concerns QA / Demo / Install databases (and potentially customer Test / Accept / Production environments)
- Add a new 'ALOpsUpgrade' task: This task would check all artifacts (app.json) files to deploy and detect if there's a higher application / platform version than the ones present on the target environment.
- If the above condition is met, [Technical upgrade](https://docs.microsoft.com/nl-be/dynamics365/business-central/dev-itpro/upgrade/upgrade-technical-upgrade-v16-v17) of the database should be performed, by uninstalling all symbols and then the container should be replaced / recreated (keeping the connection with the external database, either via ALOps step settings or an external script to recreate the container). 
- [Application Upgrade](https://docs.microsoft.com/nl-be/dynamics365/business-central/dev-itpro/upgrade/upgrade-unmodified-application-v16-to-v17) should be performed by publishing / installing the newest MS symbols (system / application / test or any other installed MS apps) taken from the bc.artifacts cache.
- Continue the CD flow, e.g with the ALOpsPublish step, that will now no longer fail, since dependencies are met.

Some remarks / restrictions:
- All of the above can be scripted but we want to make use of the ALOps / CRS expertise instead of maintaining our own scripts.
- You could focus on single tenant only / non-modified base app upgrades / upgrade paths supported by the compatibility matrix only. 
- How to deal with new apps (such as the MS Email Connectors)?

I've seen you're moving customer environments to support Docker (even for production), so it would make sense that deploying updates (not only apps) would also include platform updates and BC apps (for our customer, there's no real distinction on the BC / ISV apps)

We don't want to spent time on creating custom scripts, knowing that ALOps (CRS / iFacto) has already the expertise or with the possibility that MS (Freddy) is supporting this in a near future (you'll potentially have better insights on their roadmap)

**Comment by waldo1001** [comment]
Hi @fvet,

There are a few things that hold us back doing so:
- Microsoft doesn't make this upgrade process consistent among all support BC versions
- Many different ways to deploy BC/Apps: code customized BC, hybrid, just apps, ..
- The risk of losing data
- There is a lot of infrastructure that has to be managed as well: new NST, multiple NST, multitenancy, .. .
- different ways to of deployment: (not) Docker, Multiple server/one server, ... where to run which script?
- failed upgrades means failed production environments: it's highly important anyone who performs the upgrade, would be able to fix any step in the upgrade (or rollback).
- ...

We have lots of experience in upgrades indeed.  But this experience tells us that we haven't been able to come up with a "just works" way for deploying an upgrade.. .  You need a massively consistent deployment scenario .. which on its turn, needs to be consistent for all partners.  I don't see this possible (or happening) anytime soon.. .

Long story short - we can't put something like this on the roadmap .. .
